File: 0100-same-tree/0100-same-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // Condition 1 : Root Node dono ka khali agar hai 
        if(p==nullptr && q==nullptr){
            return true;
        }
        // orr ek null hua toh 
        if(p==nullptr || q==nullptr){
            return false;
        }
        if(p->val!=q->val){
            //agr root wley ka value dono tree mein same nhi hai 
            return false;
        }

        // Left child of the tree k liye same condition karlo 
        bool r1=isSameTree(p->left,q->left);

        // Right child of the tree k liye same condition chk karlo 
        bool r2=isSameTree(p->right,q->right);

        if(r1==true && r2==true){
            return true ;               // agar childs k liye bhi sarey condn hold huye toh true 
        }
        return false;                  // otherwise false 
    }
};

File: 0100-same-tree/0100-same-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        // agr dono hi root wley null hai 
        if(p==null && q==null){
            return true;            // null hua toh bhi same hi hai 
        }
        // agr ek koi null hua toh false 
        if(p==null || q==null){
            return false;           
        }

        // agar root wla null nhi hai ; toh value se dekhna hoga 
        if(p.val!=q.val){
            return false;
        }

        // Left child k liye bhi same conditions lena hai 
        boolean r1=isSameTree(p.left,q.left);

        // Right child k liye bhi dekhlo hold krta hai ki nhi 
        boolean r2=isSameTree(p.right,q.right);

        // Ab chk krlo ki same hai ki nhi hai right and left bhi
        if(r1==true && r2==true){
            return true;
        }
        return false;            // otherwise wla kaam
    }
}